nixos/modules/services/web-apps/coder.nix

```diff
@@ -0,0 +1,165 @@
+        default = null;
+        description = lib.mdDoc ''
+          Name of the nginx virtualhost to use and setup. If null, do not setup any virtualhost.
+        '';
+      };
+
+      listenAddress = mkOption {
+        type = types.str;
+        description = lib.mdDoc ''
+          Listen address
+        '';
+        default = "127.0.0.1:3000";
+      };
+
+      accessUrl = mkOption {
+        type = types.nullOr types.str;
+        description = lib.mdDoc ''
+          Access URL should be a external IP address or domain with DNS records pointing to Coder
+        '';
+        default = null;
+        example = "https://coder.example.com";
+      };
+
+      wildcardAccessUrl = mkOption {
+        type = types.nullOr types.str;
+        description = lib.mdDoc ''
+          If you are providing TLS certificates directly to the Coder server, you must use a single certificate for the root and wildcard domains
+        '';
+        default = null;
+        example = "*.coder.example.com";
+      };
+
+      postgresqlUrl = mkOption {
+        type = types.nullOr types.str;
+        description = lib.mdDoc ''
+          Coder uses a PostgreSQL database to store users, workspace metadata, and other deployment information
+        '';
+        default = null;
+        example = "postgresql://root:root@database/coder?sslmode=disable";
+      };
+
+      tlsCert = mkOption {
+        type = types.nullOr types.str;
+        description = lib.mdDoc ''
+          The path to the TLS certificate.
+        '';
+        default = null;
+      };
+
+      tlsKey = mkOption {
+        type = types.nullOr types.str;
+        description = lib.mdDoc ''
+          The path to the TLS key.
+        '';
+        default = null;
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.coder = let
+      # see https://github.com/coder/coder/issues/4731
+      pgCtl = pkgs.writeScript "pg_ctl" ''
+        #!${pkgs.bash}/bin/bash
+
+        exec ${pkgs.postgresql_14}/bin/pg_ctl "$@" -o '--unix_socket_directories=${cfg.homeDir}'
+      '';
+    in {
+      description = "Coder - Self-hosted developer workspaces on your infra";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        mkdir -p ${cfg.homeDir}/.config/coderv2/postgres/bin/bin
+        rm -f ${cfg.homeDir}/.config/coderv2/postgres/bin/bin/initdb
+        rm -f ${cfg.homeDir}/.config/coderv2/postgres/bin/bin/pg_ctl
+        rm -f ${cfg.homeDir}/.config/coderv2/postgres/bin/bin/postgres
+        rm -f ${cfg.homeDir}/.config/coderv2/postgres/bin/share
+        ln -s ${pkgs.postgresql_14}/bin/initdb ${cfg.homeDir}/.config/coderv2/postgres/bin/bin/initdb
+        ln -s ${pgCtl} ${cfg.homeDir}/.config/coderv2/postgres/bin/bin/pg_ctl
+        ln -s ${pkgs.postgresql_14}/bin/postgres ${cfg.homeDir}/.config/coderv2/postgres/bin/bin/postgres
+        ln -s ${pkgs.postgresql_14}/share ${cfg.homeDir}/.config/coderv2/postgres/bin/
```

Comment:
First of all, we shouldn't pin Postgres version here. 
Secondly, why is this done? We shouldn't integrate Postgres support here directly. We must provide an option that has a connection setting, and it should be up to the user how to deal with it.

Reply:
I will remove the internal postgres support